### PR TITLE
Begun work to support native file dialogues

### PR DESCRIPTION
**Pull request by OsiJr**

### CODE REVIEW

File: src/imgui_ex/vcImGuiSimpleWidgets.cpp [hunk 1]
```diff
@@ -84,6 +85,40 @@ bool vcIGSW_InputTextWithResize(const char *pLabel, char **ppBuffer, size_t *pBu
   return retVal;
 }
 
+void vcIGSW_FilePicker(vcState *pProgramState, const char *pLabel, char *pBuffer, size_t bufferSize, const char **ppExtensions, size_t numExtensions, bool read, vcFileDialogCallback onChange)
+{
+  const float ButtonWidth = 20.f;
```
Comment (Jossos) [inline src/imgui_ex/vcImGuiSimpleWidgets.cpp]: camelCase

Comment (OsiJr) [reply src/imgui_ex/vcImGuiSimpleWidgets.cpp]: Constants are proper case

File: src/main.cpp [hunk 2]
```diff
@@ -192,7 +195,7 @@ void vcMain_MainLoop(void *pArgs)
 {
   vcState *pProgramState = (vcState*)pArgs;
 #else
-void vcMain_MainLoop(vcState *pProgramState)
+void vcMain_MainLoop(vcState * pProgramState)
```
Comment (Jossos) [inline src/main.cpp]: asterisk spacing

File: src/main.cpp [hunk 3]
```diff
@@ -568,7 +571,7 @@ void vcMain_MainLoop(vcState *pProgramState)
 }
 
 #if UDPLATFORM_EMSCRIPTEN
-void vcMain_GetScreenResolution(vcState *pProgramState)
+void vcMain_GetScreenResolution(vcState * pProgramState)
```
Comment (Jossos) [inline src/main.cpp]: asterisk spacing

File: src/imgui_ex/vcImGuiSimpleWidgets.cpp [hunk 4]
```diff
@@ -84,6 +85,40 @@ bool vcIGSW_InputTextWithResize(const char *pLabel, char **ppBuffer, size_t *pBu
   return retVal;
 }
 
+void vcIGSW_FilePicker(vcState *pProgramState, const char *pLabel, char *pBuffer, size_t bufferSize, const char **ppExtensions, size_t numExtensions, bool read, vcFileDialogCallback onChange)
```
Comment (peterAdams1014) [inline src/imgui_ex/vcImGuiSimpleWidgets.cpp]: Consider renaming the `read` parameter to be `loadOnly` - more consistent with the subsequent functions.

File: src/main.cpp (hunk 2 above)
Comment (peterAdams1014) [inline src/main.cpp]: Unnecessary white space. Was this a misconfigured VS?

File: src/main.cpp [hunk 5]
```diff
@@ -2313,4 +2320,124 @@ void vcRenderWindow(vcState *pProgramState)
...
+    if (pProgramState->fileDialog.showDialog)
+    {
+      HRESULT hr;
+      IFileDialog *pFileOpen;
+
+      // Create the FileOpenDialog object.
+      if (pProgramState->fileDialog.loadOnly)
+        hr = CoCreateInstance(CLSID_FileOpenDialog, NULL, CLSCTX_ALL, IID_IFileOpenDialog, reinterpret_cast<void**>(&pFileOpen));
+      else
+        hr = CoCreateInstance(CLSID_FileSaveDialog, NULL, CLSCTX_ALL, IID_IFileSaveDialog, reinterpret_cast<void**>(&pFileOpen));
```
Comment (Jossos) [inline src/main.cpp]: HRESULT hr = CoCreateInstance(pProgramState->fileDialog.loadOnly ? CLSID_FileOpenDialog : CLSID_FileSaveDialog, NULL, CLSCTX_ALL, IID_IFileOpenDialog, reinterpret_cast<void**>(&pFileOpen));

File: src/main.cpp [hunk 6]
```diff
@@ -2313,4 +2320,124 @@ void vcRenderWindow(vcState *pProgramState)
...
   vcSettingsUI_Show(pProgramState);
   vcModals_DrawModals(pProgramState);
+
+#if UDPLATFORM_WINDOWS
+  if (pProgramState->settings.window.useNativeUI)
+  {
+    if (pProgramState->fileDialog.showDialog)
+    {
+      HRESULT hr;
+      IFileDialog *pFileOpen;
```
Comment (peterAdams1014) [inline src/main.cpp]: Uninitialized memory.

File: src/main.cpp [hunk 7]
```diff
@@ -2313,4 +2320,124 @@ void vcRenderWindow(vcState *pProgramState)
...
+
+      // Create the FileOpenDialog object.
+      if (pProgramState->fileDialog.loadOnly)
+        hr = CoCreateInstance(CLSID_FileOpenDialog, NULL, CLSCTX_ALL, IID_IFileOpenDialog, reinterpret_cast<void**>(&pFileOpen));
+      else
+        hr = CoCreateInstance(CLSID_FileSaveDialog, NULL, CLSCTX_ALL, IID_IFileSaveDialog, reinterpret_cast<void**>(&pFileOpen));
+
+      if (SUCCEEDED(hr))
+      {
+        char extBuffer[1024];
```
Comment (peterAdams1014) [inline src/main.cpp]: Uninitialized memory.

File: src/main.cpp [hunk 8]
```diff
@@ -2313,4 +2320,124 @@ void vcRenderWindow(vcState *pProgramState)
...
+
+        // Get the file name from the dialog box.
+        if (SUCCEEDED(hr))
+        {
+          // Show the Open dialog box.
+          hr = pFileOpen->Show(NULL);
+
+          if (SUCCEEDED(hr))
+          {
+            IShellItem *pItem;
```
Comment (peterAdams1014) [inline src/main.cpp]: Uninitialized memory.

File: src/main.cpp [hunk 9]
```diff
@@ -2313,4 +2320,124 @@ void vcRenderWindow(vcState *pProgramState)
...
+            hr = pFileOpen->GetResult(&pItem);
+
+            if (SUCCEEDED(hr))
+            {
+              PWSTR pszFilePath;
```
Comment (peterAdams1014) [inline src/main.cpp]: Uninitialized memory.

File: src/vcConvert.cpp [hunk 10]
```diff
@@ -518,6 +516,7 @@ void vcConvert_ShowUI(vcState *pProgramState)
       if (ImGui::InputText("##vcSetTemporaryDirectoryText", tempDirectory, udLengthOf(tempDirectory)))
         vdkConvert_SetTempDirectory(pSelectedJob->pConvertContext, tempDirectory);
 
+      /*
```
Comment (Jossos) [inline src/vcConvert.cpp]: could you either remove the contents within this comment, or put a note describing why it's been commented out and left here

File: src/vcModals.cpp [hunk 11]
```diff
@@ -340,51 +315,23 @@ void vcModals_DrawFileModal(vcState *pProgramState)
...
 
     if (loadFile)
     {
 #if VC_HASCONVERT
-      if (mode == vcMT_ConvertAdd)
-      {
-        vcConvert_QueueFile(pProgramState, pProgramState->modelPath);
-      }
-      else if (mode == vcMT_ConvertOutput)
+      /*if (mode == vcMT_ConvertOutput)
```
Comment (peterAdams1014) [inline src/vcModals.cpp]: Why is this commented out?

Comment (Jossos) [inline src/vcModals.cpp]: remove, or add an explanation for comment please

File: src/imgui_ex/vcFileDialog.h [hunk 12]
```diff
@@ -2,7 +2,65 @@
...
+static const char *SupportedFileTypes_Images[] = { ".jpg", ".png", ".tga", ".bmp", ".gif" };
+static const char *SupportedFileTypes_Projects[] = { ".json", ".udp" };
+static const char *SupportedFileTypes_SceneItems[] = { ".uds", ".ssf", ".udg" };
+
+static const char *SupportedTileTypes_ConvertExport[] = { ".uds" };
+static const char *SupportedFileTypes_ConvertImport[] = { ".uds", ".ssf", ".udg", ".las", ".laz", ".e57", ".txt", ".csv", ".asc", ".pts", ".ptx", ".xyz", ".obj"
+#ifdef FBXSDK_ON
+  , ".fbx"
+#endif
+};
```
[the remark this answers is not in the document]
Comment (samsinsane) [reply src/imgui_ex/vcFileDialog.h]: I think the benefits of having these in the header outweigh the benefits of having them in a cpp. These variables will be passed into the template functions below and have their size automatically determined which reduces the chances of making mistakes.

File: src/imgui_ex/vcFileDialog.h [hunk 13]
```diff
@@ -2,7 +2,65 @@
...
+  bool showDialog;
+  bool folderOnly;
+  bool loadOnly;
+  char *pPath;
+  size_t pathLen;
+
+  const char **ppExtensions; // Should be static arrays only
+  size_t numExtensions;
+
+  vcFileDialogCallback onSelect;
```
[the remark this answers is not in the document]
Comment (samsinsane) [reply src/imgui_ex/vcFileDialog.h]: We don't set default values for struct members, that would create a constructor that won't be called when using `udAllocType`.

File: src/imgui_ex/vcFileDialog.h [hunk 14]
```diff
@@ -2,7 +2,65 @@
...
+  pDialog->ppExtensions = ppExtensions;
+  pDialog->numExtensions = numExtensions;
+
+  pDialog->onSelect = callback;
+}
+
+template <size_t N, size_t M> inline void vcFileDialog_Show(vcFileDialog *pDialog, char(&path)[N], const char *(&extensions)[M], bool loadOnly, vcFileDialogCallback callback)
+{
+  vcFileDialog_Show(pDialog, path, N, extensions, M, loadOnly, callback);
+}
```
[the remark this answers is not in the document]
Comment (samsinsane) [reply src/imgui_ex/vcFileDialog.h]: This is a templated function and must be in the header.

File: src/vcModals.cpp [hunk 15]
```diff
@@ -340,51 +315,23 @@ void vcModals_DrawFileModal(vcState *pProgramState)
...
-        loadFile = true;
+      vcIGSW_FilePicker(pProgramState, "Filename", pProgramState->modelPath, SupportedFileTypes_SceneItems, true, [pProgramState] {
+        pProgramState->loadList.PushBack(udStrdup(pProgramState->modelPath));
+      });
     }
     else if (mode == vcMT_ExportProject)
     {
       const char *fileExtensions[] = { ".json" };
-      if (vcFileDialog_Show(pProgramState->modelPath, sizeof(pProgramState->modelPath), false, fileExtensions, udLengthOf(fileExtensions)))
+      if (vcFileDialog_DrawImGui(pProgramState->modelPath, sizeof(pProgramState->modelPath), false, fileExtensions, udLengthOf(fileExtensions)))
```
Comment (samsinsane) [inline src/vcModals.cpp]: Isn't there a template helper to avoid these `sizeof` and `udLengthOf` calls?

File: src/vcSettingsUI.cpp [hunk 16]
```diff
@@ -449,45 +449,60 @@ void vcSettingsUI_Show(vcState *pProgramState)
...
-              vcTexture_Destroy(&pProgramState->settings.convertdefaults.watermark.pTexture);
+              // TODO: Resize watermark to the same dimensions as vdkConvert does - maybe requires additional VDK functionality?
+              filename.SetFolder(pProgramState->settings.pSaveFilePath);
+              udFile_Save(filename, pData, (size_t)dataLength);
+              udFree(pData);
             }
-          }
+            udStrcpy(pProgramState->settings.convertdefaults.watermark.filename, filename.GetFilenameWithExt());
+            pProgramState->settings.convertdefaults.watermark.isDirty = true;
+            });
```
Comment (samsinsane) [inline src/vcSettingsUI.cpp]: Indentation on this line is wrong.

File: src/imgui_ex/vcImGuiSimpleWidgets.cpp (hunk 4 above)
Comment (OsiJr) [reply src/imgui_ex/vcImGuiSimpleWidgets.cpp]: This was changed to `read` because `loadOnly` is wrong.

File: src/main.cpp (hunk 5 above)
Comment (OsiJr) [reply src/main.cpp]: There are 2 variables that are changed here so rather than double ternary I split it out

File: src/vcModals.cpp (hunk 15 above)
Comment (peterAdams1014) [reply src/vcModals.cpp]: There is not.

**Pull request merged**